### ISSUE TITLE
Linux encrypted home: TypeError for any URL above 104 characters

### ISSUE THREAD

**Issue by nicolas-raoul** [sindresorhus/pageres#154]
Any URL above 104 characters makes pageres choke immediately (before even sending any web request it seems):

```
$ pageres "https://tools.wmflabs.org/wikivoyage/w/poimap2.php?lat=42.333333&lon=12.283333&zoom=13&layer=O&lang=en&name=Fabrica_di_Roma" 1680x1080

events.js:74
        throw TypeError('Uncaught, unspecified "error" event.');
              ^
TypeError: Uncaught, unspecified "error" event.
    at TypeError (<anonymous>)
    at EventEmitter.emit (events.js:74:15)
    at cleanupSync (/usr/local/lib/node_modules/pageres/node_modules/fs-write-stream-atomic/index.js:46:42)
    at WriteStream.emit (/usr/local/lib/node_modules/pageres/node_modules/fs-write-stream-atomic/index.js:53:12)
    at WriteStream.<anonymous> (evalmachine.<anonymous>:1660:12)
    at OpenReq.Req.done (/usr/local/lib/node_modules/pageres/node_modules/fs-write-stream-atomic/node_modules/graceful-fs/graceful-fs.js:141:5)
    at OpenReq.done (/usr/local/lib/node_modules/pageres/node_modules/fs-write-stream-atomic/node_modules/graceful-fs/graceful-fs.js:61:22)
    at Object.oncomplete (evalmachine.<anonymous>:107:15)
```
pageres 1.1.0
Ubuntu 2014.04
nodejs v0.10.25
http_parser: '1.0'
node: '0.10.25'
v8: '3.14.5.9'
ares: '1.10.0'
uv: '0.10.23'
zlib: '1.2.8'
modules: '11'
openssl: '1.0.1f'
npm: '1.3.10'

Difference between URLs of respectively 105 and 104 characters:
```
$ pageres "http://123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123.com" 1680x1080

events.js:74
        throw TypeError('Uncaught, unspecified "error" event.');
              ^
TypeError: Uncaught, unspecified "error" event.
    at TypeError (<anonymous>)
    at EventEmitter.emit (events.js:74:15)
    at cleanupSync (/usr/local/lib/node_modules/pageres/node_modules/fs-write-stream-atomic/index.js:46:42)
    at WriteStream.emit (/usr/local/lib/node_modules/pageres/node_modules/fs-write-stream-atomic/index.js:53:12)
    at WriteStream.<anonymous> (evalmachine.<anonymous>:1660:12)
    at OpenReq.Req.done (/usr/local/lib/node_modules/pageres/node_modules/fs-write-stream-atomic/node_modules/graceful-fs/graceful-fs.js:141:5)
    at OpenReq.done (/usr/local/lib/node_modules/pageres/node_modules/fs-write-stream-atomic/node_modules/graceful-fs/graceful-fs.js:61:22)
    at Object.oncomplete (evalmachine.<anonymous>:107:15)

$ pageres "http://12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012.com" 1680x1080
Couldn't load url: http://12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012.com
```

**Comment by sindresorhus** [comment]
Seems to work for me on OS X:

```
❯ pageres "https://tools.wmflabs.org/wikivoyage/w/poimap2.php?lat=42.333333&lon=12.283333&zoom=13&layer=O&lang=en&name=Fabrica_di_Roma" 1680x1080

✔ Generated 1 screenshot from 1 url and 1 resolution
```

**Comment by kevva** [comment]
Just we need to catch errors from the `write` stream too https://github.com/sindresorhus/pageres/blob/master/lib/util.js#L109? Thought that was handled by listening for errors on `pipe`.

**Comment by sindresorhus** [comment]
@kevva You gotta learn. In programming nothing is handled for you :p

**Comment by kevva** [comment]
At least not when it comes to node streams.

**Comment by kevva** [comment]
Works fine for me too on Ubuntu.

```sh
$ pageres "https://tools.wmflabs.org/wikivoyage/w/poimap2.php?lat=42.333333&lon=12.283333&zoom=13&layer=O&lang=en&name=Fabrica_di_Roma" 1680x1080

✔ Generated 1 screenshot from 1 url and 1 resolution
```

**Comment by nicolas-raoul** [comment]
@kevva What Ubuntu version?
I am using the latest LTS version, which is 2014.04

**Comment by kevva** [comment]
14.10.

**Comment by nicolas-raoul** [comment]
For others who might have the same problem, a low-tech workaround is to first generate a `frameset.html` page like this:
```
<frameset rows="100%">
  <frameset cols="100%">
    <frame src="https://tools.wmflabs.org/wikivoyage/w/poimap2.php?lat=42.333333&lon=12.283333&zoom=13&layer=O&lang=en&name=Fabrica_di_Roma" frameborder="0" scrolling="no">
  </frameset>
</frameset>
```
and then take a screenshot like this:
```
pageres file:///path/to/frameset.html
```

**Comment by nicolas-raoul** [comment]
Sorry I was wrong, the workaround does not work, same error :-(

**Comment by nicolas-raoul** [comment]
If there is anything I can do to debug the problem, I would be glad to help.
I installed pageres exactly as specified in the README.

**Comment by sindresorhus** [comment]
@nicolas-raoul We just need to handle the error and then have you run it to see what it says. We'll let you know.

Quite weird though. Seems like atom write throws when the filename get's very long, but only on @nicolas-raoul's system.

**Comment by nicolas-raoul** [comment]
By any chance, are you generating a `server!very!long!path.png` filename and then renaming it to the value of `--filename`?

As you can see in the five first lines below, my Ubuntu limits file names to 144 characters in `$HOME` and 255 characters in `/tmp`. The difference is due to Ubuntu's home encryption layer.
Anyway, I run pageres in `/tmp` so home encryption is not a valid reason to crash.

```
nico@nico:~$ >`printf 'a%.0s' {1..143}`
nico@nico:~$ >`printf 'a%.0s' {1..144}`
bash: `printf 'a%.0s' {1..144}`: File name too long
nico@nico:~$ cd /tmp
nico@nico:/tmp$ >`printf 'a%.0s' {1..255}`
nico@nico:/tmp$ >`printf 'a%.0s' {1..256}`
bash: `printf 'a%.0s' {1..256}`: File name too long
nico@nico:/tmp$ pageres --filename /tmp/out.png "https://tools.wmflabs.org/wikivoyage/w/poimap2.php?lat=42.333333&lon=12.283333&zoom=13&layer=O&lang=en&name=Fabrica_di_Roma" 1680x1080

events.js:74
        throw TypeError('Uncaught, unspecified "error" event.');
              ^
TypeError: Uncaught, unspecified "error" event.
    at TypeError (<anonymous>)
    at EventEmitter.emit (events.js:74:15)
    at cleanupSync (/usr/local/lib/node_modules/pageres/node_modules/fs-write-stream-atomic/index.js:46:42)
    at WriteStream.emit (/usr/local/lib/node_modules/pageres/node_modules/fs-write-stream-atomic/index.js:53:12)
    at WriteStream.<anonymous> (evalmachine.<anonymous>:1660:12)
    at OpenReq.Req.done (/usr/local/lib/node_modules/pageres/node_modules/fs-write-stream-atomic/node_modules/graceful-fs/graceful-fs.js:141:5)
    at OpenReq.done (/usr/local/lib/node_modules/pageres/node_modules/fs-write-stream-atomic/node_modules/graceful-fs/graceful-fs.js:61:22)
    at Object.oncomplete (evalmachine.<anonymous>:107:15)
```

I humbly suggest you use short temporary names in `/tmp` for all temporary filenames (with `mktemp` or similar).

**Comment by kevva** [comment]
@nicolas-raoul, it's [fs-write-stream-atomic](https://github.com/npm/fs-write-stream-atomic) that's adding a bunch of stuff to the filename.

**Comment by kevva** [comment]
Haven't been able to reproduce this but I've added error handling for the atomic write-stream in https://github.com/sindresorhus/pageres/commit/71e5979be5f0f0f607be04696331fefaafa3c8d5. Going to close.

**Comment by nicolas-raoul** [comment]
Bug still present in pageres 1.3.0, please reopen:

```
$ pageres --version
1.3.0
$ pageres "https://tools.wmflabs.org/wikivoyage/w/poimap2.php?lat=42.333333&lon=12.283333&zoom=13&layer=O&lang=en&name=Fabrica_di_Roma" 1680x1080

events.js:74
        throw TypeError('Uncaught, unspecified "error" event.');
              ^
TypeError: Uncaught, unspecified "error" event.
    at TypeError (<anonymous>)
    at EventEmitter.emit (events.js:74:15)
    at cleanupSync (/usr/local/lib/node_modules/pageres/node_modules/fs-write-stream-atomic/index.js:46:42)
    at WriteStream.emit (/usr/local/lib/node_modules/pageres/node_modules/fs-write-stream-atomic/index.js:53:12)
    at WriteStream.<anonymous> (evalmachine.<anonymous>:1660:12)
    at OpenReq.Req.done (/usr/local/lib/node_modules/pageres/node_modules/fs-write-stream-atomic/node_modules/graceful-fs/graceful-fs.js:141:5)
    at OpenReq.done (/usr/local/lib/node_modules/pageres/node_modules/fs-write-stream-atomic/node_modules/graceful-fs/graceful-fs.js:61:22)
    at Object.oncomplete (evalmachine.<anonymous>:107:15)
```

**Comment by sindresorhus** [comment]
@kevva Looks like we're still not catching all the stream errors.

**Comment by kevva** [comment]
@sindresorhus, I think it's fs-write-stream-atomic that's not catching all the errors.

**Comment by nicolas-raoul** [comment]
I updated to Ubuntu 2014.10/pageres 1.4.0 and now I get a different error:
```
$ pageres "https://tools.wmflabs.org/wikivoyage/w/poimap2.php?lat=42.333333&lon=12.283333&zoom=13&layer=O&lang=en&name=Fabrica_di_Roma" 1680x1080

/usr/local/lib/node_modules/pageres/cli.js:98
				throw err;
				      ^
Error: ENAMETOOLONG, open '/home/nico/tools.wmflabs.org!wikivoyage!w!poimap2.php!lang=en&lat=42.333333&layer=O&lon=12.283333&name=Fabrica_-1680x1080.png.3f6f08b6d459421985fc3619a713980f'
```

**Comment by nicolas-raoul** [comment]
Same error as immediately above on Ubuntu 2015.04

I have found a workaround: Do NOT use a encrypted home, and that will work.

Anyone wanting to using pageres while staying secure must create another user dedicated to pagers execution.

**Comment by nicolas-raoul** [comment]
Fixed with Pageres 3.0.2 and NodeJS 5.2.0 on Ubuntu 2015.04 encrypted home, thanks!
